### PR TITLE
DEP: Remove deprecated numeric types and deprecate remaining

### PR DESCRIPTION
**Pull request by seberg**
This removes the types from usage for `np.dtype()`, since the
two are closely related, also finishes the deprecation of
`typeNA` and `sctypeNA` (thus removing them from the public API).

The deprecation for the numeric types was only shown for
`np.dtype("Complex64")`, etc. however, here they are also
removed from `np.typeDict` and `np.sctypeDict`.

### CODE REVIEW

File: numpy/core/_type_aliases.py [hunk 1]
```diff
@@ -127,27 +106,20 @@ def _add_aliases():
...
-        sctypeNA[na_name] = info.type
-        sctypeNA[info.type] = na_name
-        sctypeNA[info.char] = na_name
-
         sctypeDict[char] = info.type
-        sctypeNA[char] = na_name
+
+    # Add deprecated numeric-style type aliases manually, at some point
+    # we may want to deprecate the lower case "bytes0" version as well.
+    for name in ["Bytes0", "Datetime64", "Str0", "Uint64"]:
```
Comment (eric-wieser) [inline numpy/core/_type_aliases.py:119]: Why specifically these ones?

Comment (seberg) [reply numpy/core/_type_aliases.py:119]: I looked at the diff between before and after these changes. And those were all the remaining ones, so I figured we might as well deprecate them as well... It seems weird to me to get rid of Timedelta, but keep Datetime.

**Comment by seberg** [comment]
For everyone, the difference between `typeDict` before and after this branch is:
```
{'Bool',  'Complex128', 'Complex32', 'Complex64', 'Float128', 'Float16', 'Float32', 'Float64',
 'Int16', 'Int32', 'Int64', 'Int8', 'Object0', 'Timedelta64', 'UInt16', 'UInt32', 'UInt64', 'UInt8',
 'Void0'}
```
You see that `Datetime64`, etc. are the ones I explicitly added back. And on this branch, these are the only (more than single characters or "M8") type codes which start with an upper case letter:
```
[...
 'Bytes0',
 'Datetime64',
 'Str0',
 'Uint64']
```
which is the ones I added to be deprecated now.

File: numpy/core/_type_aliases.py [hunk 2]
```diff
@@ -127,27 +106,23 @@ def _add_aliases():
...
+        if name == "Uint64" and "uint64" not in allTypes:
+            # Due to the integer priority, this was not always defined
+            continue
```
Comment (seberg) [inline numpy/core/_type_aliases.py:122]: This if is weird, but it failed tests and I think it is correct... if anyone has a 32bit linux or maybe windows to see that `Uint64` was never included in `np.typeDict` on old versions, that would be good.

Comment (mattip) [reply numpy/core/_type_aliases.py:122]: It now passes tests, is this comment outdated?

Comment (seberg) [reply numpy/core/_type_aliases.py:122]: No, I worked around it. But, I now realized on windows we probably had a stray `Uint32` and that will be missing now...

Comment (mattip) [reply numpy/core/_type_aliases.py:122]: np.typeDict has `UInt64` on 1.16, windows 32 bit:
```
>>> sys.version
'2.7.18 (v2.7.18:8d21aa21f2, Apr 20 2020, 13:19:08) [MSC v.1500 32 bit (Intel)]'
>>> np.typeDict['UInt64']
<type 'numpy.uint64'>
>>> np.__version__
1.16.6
>>>
```

Comment (mattip) [reply numpy/core/_type_aliases.py:122]: note the 'I' in 'UInt64'

File: doc/release/upcoming_changes/16554.compatibility.rst [hunk 3]
```diff
@@ -0,0 +1,10 @@
+Numeric style types removed from type dictionaries
+--------------------------------------------------
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16554.compatibility.rst:2]:
```suggestion
Numeric-style type names have been removed from type dictionaries
-----------------------------------------------------------------
```

File: doc/release/upcoming_changes/16554.deprecation.rst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Further Numeric Style types Deprecated
+--------------------------------------
+
+The remaining numeric-style type codes ``"Bytes0"``,
+``Str0``, ``Uint64``, and ``Datetime64`` have been
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16554.deprecation.rst:5]:
```suggestion
The remaining numeric-style type codes ``Bytes0``,
``Str0``, ``Uint64``, and ``Datetime64`` have been
```

File: doc/release/upcoming_changes/16554.expired.rst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+* The deprecation of numeric style type-codes ``np.dtype("Complex64")``
+  (with upper case spelling), is expired.  ``"Complex64"`` corresponds to
+  ``"complex128"`` and ``"Complex32"`` corresponds to ``"complex64"``.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/16554.expired.rst:3]:
```suggestion
  (with upper case spelling), is expired.  ``"Complex64"`` corresponded to
  ``"complex128"`` and ``"Complex32"`` corresponded to ``"complex64"``.
```

File: numpy/core/_type_aliases.py [hunk 6]
```diff
@@ -11,40 +11,19 @@
...
 .. data:: sctypes
     A dictionary keyed by a "type group" string, providing a list of types
     under that group.
 
 """
-import warnings
 
 from numpy.compat import unicode
-from numpy._globals import VisibleDeprecationWarning
-from numpy.core._string_helpers import english_lower, english_capitalize
```
Comment (eric-wieser) [inline numpy/core/_type_aliases.py:29]: Is `english_capitalize` used any more?

File: numpy/core/src/multiarray/descriptor.c [hunk 7]
```diff
@@ -1678,14 +1678,13 @@ _convert_from_str(PyObject *obj, int align)
         }
 
         /* Check for a deprecated Numeric-style typecode */
-        char *dep_tps[] = {"Bool", "Complex", "Float", "Int",
-                           "Object0", "String0", "Timedelta64",
-                           "Unicode0", "UInt", "Void0"};
+        char *dep_tps[] = {"Bytes", "Datetime64", "Str", "Uint64"};
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c:1681]: Perhaps to play it safe:
```suggestion
        /* `Uint` has deliberately weird uppercasing */
        char *dep_tps[] = {"Bytes", "Datetime64", "Str", "Uint"};
```

File: numpy/core/_type_aliases.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/_type_aliases.py:122]: The capitalized `UInt` form has already been deprecated - it's just the weird `Uint` form that isn't.

It looks to me like I added this type accidentally as part of giving `[np.cint, np.int_, np.longlong]` unique names.  On numpy 1.11.0 (and python 2.7) the name isn't there at all:
```python
>>> numpy.__version__
'1.11.0'
>>> sorted(k for k in numpy.typeDict.keys() if isinstance(k, str) and k[0].isupper() and len(k) > 2)
['Bool', 'Complex128', 'Complex32', 'Complex64', 'Datetime64', 'Float128', 'Float16', 'Float32', 'Float64', 'Int16', 'Int32', 'Int64', 'Int8', 'Object0', 'String0', 'Timedelta64', 'UInt16', 'UInt32', 'UInt64', 'UInt8', 'Unicode0', 'Void0']
```

So I think we can just remove this special case?

**Review by eric-wieser — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks, will these changes end up in NumPy 1.19 or 1.20?

**Comment by seberg** [comment]
1.20, no intention to backport this, 1.19 may still get more critical fixes, this is not critical.

**Pull request merged**